### PR TITLE
Create pattern

### PR DESCRIPTION
**Pull request by Magisus**
Allows the user to create a new pattern and save it in the database. It can then be viewed as a normal pattern.

Editing and deleting patterns will be handled in a future pull request, as will styling refinements.

### CODE REVIEW

File: app/src/main/java/com/connorlay/knitgrid/activities/PatternCreationActivity.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            buttonBar.addView(button, 70, 70);
+        }
+        buttonBar.setVisibility(View.GONE);
+    }
+
+    @OnClick(R.id.save_pattern_button)
+    public void savePattern() {
+        PatternDetailFragment frag = (PatternDetailFragment)
+                getSupportFragmentManager().findFragmentById(R.id.pattern_detail_frame);
+        if (frag != null) {
```
Comment (connorlay) [inline app/src/main/java/com/connorlay/knitgrid/activities/PatternCreationActivity.java]: I think this is a great place to use a guard statement.

``` Java
if (frag == null) {
     return;
}
// rest of code here
```

Thoughts?

File: app/src/main/java/com/connorlay/knitgrid/activities/PatternCreationActivity.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        List<Stitch> stitches = Stitch.listAll(Stitch.class);
+        for (final Stitch stitch : stitches) {
+            Button button = new Button(this);
+            button.setBackground(getDrawable(stitch.getIconID()));
+            button.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    PatternDetailFragment frag = (PatternDetailFragment)
+                            getSupportFragmentManager().findFragmentById(R.id.pattern_detail_frame);
+                    if (frag != null) {
```
Comment (connorlay) [inline app/src/main/java/com/connorlay/knitgrid/activities/PatternCreationActivity.java]: Same thing about guard statements.

File: app/src/main/java/com/connorlay/knitgrid/fragments/CreatePatternDialogFragment.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        listener = (PatternCreationListener) getActivity();
+
+        View rootView = LayoutInflater.from(getActivity()).inflate(R.layout
+                .dialog_fragment_create_pattern, null);
+
+        ButterKnife.bind(this, rootView);
+
+        AlertDialog dialog = new AlertDialog.Builder(getActivity())
+                .setView(rootView)
+                .setTitle(getActivity().getString(R.string.create_pattern_dialog_title))
```
Comment (connorlay) [inline app/src/main/java/com/connorlay/knitgrid/fragments/CreatePatternDialogFragment.java]: I think you can use Butterknife to bind strings.

File: app/src/main/java/com/connorlay/knitgrid/fragments/CreatePatternDialogFragment.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                .setPositiveButton(getActivity().getString(R.string
+                                .create_pattern_dialog_positive_button),
+                        new DialogInterface.OnClickListener() {
+                            @Override
+                            public void onClick(DialogInterface dialog, int which) {
+                                String name = nameInput.getText().toString();
+                                String rowCountString = rowsInput.getText().toString();
+                                String colCountString = columnsInput.getText().toString();
+                                if (name.isEmpty()) {
```
Comment (connorlay) [inline app/src/main/java/com/connorlay/knitgrid/fragments/CreatePatternDialogFragment.java]: You could also refactor this to use guard statements.

File: app/src/main/java/com/connorlay/knitgrid/fragments/PatternDetailFragment.java [hunk 5]
```diff
@@ -25,6 +25,11 @@
  * Created by connorlay on 2/26/16.
  */
 public class PatternDetailFragment extends Fragment {
+
+    public interface RowColumnListener {
```
Comment (connorlay) [inline app/src/main/java/com/connorlay/knitgrid/fragments/PatternDetailFragment.java]: I'm not a big fan of this interface name. What about `CellSelectedListener` or something similar?

File: app/src/main/java/com/connorlay/knitgrid/fragments/PatternDetailFragment.java [hunk 6]
```diff
@@ -41,6 +46,8 @@
 
     private PatternPresenter mPatternPresenter;
 
+    private RowColumnListener mListener;
+
     public static PatternDetailFragment newInstance(PatternPresenter patternPresenter) {
```
Comment (connorlay) [inline app/src/main/java/com/connorlay/knitgrid/fragments/PatternDetailFragment.java]: Now that I think about it, we should probably be injecting a listener both for viewing and editing of patterns. Both can implement the `RowColumnListener` so there is only one constructor. It seems weird to me that we are only injecting creation behavior, but not highlighting behavior :/

**Comment by connorlay** [comment]
I left some feedback for you to consider. Overall this looks good to me and I'm excited to play with the pattern creator!

File: app/src/main/java/com/connorlay/knitgrid/fragments/PatternDetailFragment.java (hunk 6 above)
Comment (Magisus) [inline app/src/main/java/com/connorlay/knitgrid/fragments/PatternDetailFragment.java]: The issue with that is that there is nothing to inject for highlighting. If you look at the way the code works, the listener that is injected is only for communication outside of the fragment. I still have to create an OnClickListener, in parallel with highlighting. (lines 108-126) For creation, it also calls out to this listener. But that isn't how highlighting works. What were you thinking we could inject?

Comment (connorlay) [inline app/src/main/java/com/connorlay/knitgrid/fragments/PatternDetailFragment.java]: We could define some kind of interface that encapsulates cell selection behavior. That way we can unit test it and the code in general will be easier to read.

Comment (Magisus) [inline app/src/main/java/com/connorlay/knitgrid/fragments/PatternDetailFragment.java]: We should talk more about this. I'm gonna leave it as is for this PR, and then we can rethink the design after we've had a chance to discuss what it should look like.

Comment (connorlay) [inline app/src/main/java/com/connorlay/knitgrid/fragments/PatternDetailFragment.java]: Ok that sounds good to me.

**Comment by connorlay** [comment]
Looks good to me! Go ahead and squash.

**Comment by Magisus** [comment]
Squashed and ready to go!

**Pull request merged**